### PR TITLE
#336: Reworked terraform to use EC2 instead of ECS

### LINKED ISSUE #336: [Infra]: Cost optimizing AWS
### Contact details

_No response_

### Addressed problem

Currently InfraCost estimates the monthly cost to be 75$. For the usage of this project that is not reasonable cost.

### Feature description

As things stand using ECS is costing roughly 20$ for both front and backend. However, the terraform is configured such that both have their own application load balancers (ALB) that cost 20$ each. I would recommend, just forgoing ECS and moving the infra to one EC2 instance. This will be more prone to downtime as the whole infra would be reliant on one availability zone, however considering the lack of criticality this seems like a good trade off. I estimate the cost to go down to 10-20$ with this architecture. 

### Alternative solutions

The ECS could be also configured to use one ALB instead of two. This would bring the cost down to roughly 55$

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by ekosfin**
## Summary

Closes #336 This should reduce operational costs.

## Notable changes

Updated terraform to use EC2 instance instead of ECS

### CODE REVIEW

File: infra/main.tf [hunk 1]
```diff
@@ -1,354 +1,147 @@
...
-    container_port   = 8080
-  }
-
-  network_configuration {
-    subnets          = ["${aws_default_subnet.default_subnet_a.id}", "${aws_default_subnet.default_subnet_b.id}"]
-    assign_public_ip = true
-    security_groups  = ["${aws_security_group.aws-service_security_group.id}"] # Setting the security group
-  }
+resource "aws_route53_zone" "primary" {
+  name = "stagnum.com"
```
Comment (ekosfin) [inline infra/main.tf:130]: Is this domain correct?

File: infra/main.tf [hunk 2]
```diff
@@ -1,354 +1,147 @@
...
-    container_name   = "${var.app_name}-front-container"
-    container_port   = 3000 # Specifying the container port
-  }
-
-  network_configuration {
-    subnets          = ["${aws_default_subnet.default_subnet_a.id}", "${aws_default_subnet.default_subnet_b.id}"]
-    assign_public_ip = true
-    security_groups  = ["${aws_security_group.aws-service_security_group.id}"] # Setting the security group
-  }
+resource "aws_volume_attachment" "this" {
```
Comment (ekosfin) [inline infra/main.tf:119]: The postgres data should be stored here so that even if the ec2 instance would be destroyed

File: infra/main.tf [hunk 3]
```diff
@@ -1,354 +1,147 @@
...
-    path                = "/"
-    interval            = 60
-    timeout             = 10
-    healthy_threshold   = 5
-    unhealthy_threshold = 5
-    matcher             = "200-399"
-  }
+resource "aws_key_pair" "deployer" {
+  key_name   = "deployer-key"
+  public_key = "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIOgP6TjSCjZS/VWhixYYevHGdzVN4jmlT5KH9va5CiBs elias.samuli@gmail.com"
```
Comment (ekosfin) [inline infra/main.tf:94]: You might want to update this to be your public ssh key if you need access to the server

File: infra/main.tf [hunk 4]
```diff
@@ -1,354 +1,147 @@
...
+  name        = "Stagnum-security-group"
+  description = "Security group for stagnum"
+  vpc_id      = module.vpc.vpc_id
 
-# Providing a reference to our default subnets
-resource "aws_default_subnet" "default_subnet_a" {
-  availability_zone = "${var.aws_region}a"
-}
+  ingress_cidr_blocks = ["0.0.0.0/0"]
+  ingress_rules       = ["http-80-tcp", "https-443-tcp", "all-icmp", "ssh-tcp"]
```
Comment (ekosfin) [inline infra/main.tf:60]: I am still a bit unsure of what ports need to be open, so I just test the usual

File: infra/main.tf [hunk 5]
```diff
@@ -1,354 +1,147 @@
...
-data "aws_iam_policy_document" "assume_role_policy" {
-  statement {
-    actions = ["sts:AssumeRole"]
-
-    principals {
-      type        = "Service"
-      identifiers = ["ecs-tasks.amazonaws.com"]
-    }
+  data_inputs = {
+    # TODO: Set values
```
Comment (ekosfin) [inline infra/main.tf:10]: I am still a bit unsure of these env variables. Please look over them

File: infra/config/userdata.tftpl [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+echo "CUSTOM_WEIGHT_SCALE=${custom_weight_scale}" > .env
+echo "USER_WEIGHT_SCALE=${user_weight_scale}" > .env
+echo "PSEUDO_RANDOM_FLOOR=${pseudu_random_floor}" > .env
+echo "PSEUDO_RANDOM_CEILING=${pseudo_random_ceiling}" > .env
+
+echo "NEXT_PUBLIC_FRONTEND_URI=${frontend_uri}" > .env
+echo "NEXT_PUBLIC_BACKEND_URI=${backend_uri}" > .env
+echo "spotify_client_id=${spotify_client_id}" > secret_spotify_client_id.txt
+echo "spotify_client_secret=${spotify_client_secret}" > secret_spotify_client_secret.txt
+wget https://raw.githubusercontent.com/EddieTheCubeHead/Stagnum/master/docker-compose-deploy.yml #TODO: edit this to be one from config once in main
```
Comment (ekosfin) [inline infra/config/userdata.tftpl:50]: Will change this after the this PR docker compose is merged to master.

**Review by ekosfin — COMMENTED** [review]
@EddieTheCubeHead could you look though these comments.

File: infra/main.tf (hunk 1 above)
Comment (EddieTheCubeHead) [reply infra/main.tf:130]: stagnum.net is the correct one

**Review by EddieTheCubeHead — APPROVED** [review]
LGTM

**Review by EddieTheCubeHead — CHANGES_REQUESTED** [review]
Forgot before clicking approve: need to change to domain still

stagnum.net

File: infra/main.tf (hunk 1 above)
Comment (ekosfin) [reply infra/main.tf]: updated

**Review by EddieTheCubeHead — APPROVED** [review]

**Pull request merged**